pytorch_lightning/utilities/cli.py

```diff
@@ -333,8 +348,21 @@ def before_instantiate_classes(self) -> None:
     def instantiate_classes(self) -> None:
         """Instantiates the classes using settings from self.config"""
         self.config_init = self.parser.instantiate_classes(self.config)
-        self.datamodule = self.config_init.get('data')
-        self.model = self.config_init['model']
+
+        datamodule_config = self.config_init.get('data')
+        if isinstance(self.datamodule_class, type):
+            self.datamodule = datamodule_config
+        elif self.datamodule_class is not None and datamodule_config is not None:
+            self.datamodule = self.datamodule_class(**datamodule_config)
```

Comment:
@mauvilsa could `add_function_arguments` have a flag similar to `instantiate` so we avoid this?

This doesn't block this PR, just as an off comment

Reply:
I haven't looked in detail at all these changes. But I think as is now the argument linking would not work. Can this wait a bit for me to analyze it?

Reply:
Sure, we aren't targeting 1.4 for this

Reply:
I propose the following. jsonargparse would have a new function `class_from_function`. With this instead of having two cases, only `add_class_arguments` would be used and not `add_function_arguments`. At the start of `add_lightning_class_args` it would be like:

```python
if callable(lightning_class):
    lightning_class = class_from_function(lightning_class)
```

The callable would be expected to have a return type annotation such that the `issubclass` after prevents giving a function that returns an unexpected class. Since only `add_class_arguments` would be used, then `self.datamodule_class(**datamodule_config)` is not necessary and the linking of arguments would work correctly.

Reply:
Sounds good to me

Reply:
`class_from_function` is now implemented and part of jsonargparse v3.17.0.